lib/supavisor/handlers/proxy/client.ex

```diff
@@ -0,0 +1,784 @@
+        parameters: [application_name: "Supavisor auth_query"],
+        ssl: tenant.upstream_ssl,
+        socket_options: [
+          Helpers.ip_version(tenant.ip_version, tenant.db_host)
+        ],
+        queue_target: 1_000,
+        queue_interval: 5_000,
+        ssl_opts: ssl_opts || []
+      )
+
+    # kill the postgrex connection if the current process exits unexpectedly
+    Process.link(conn)
+
+    Logger.debug(
+      "ProxyClient: Connected to db #{tenant.db_host} #{tenant.db_port} #{tenant.db_database} #{user.db_user}"
+    )
+
+    resp =
+      with {:ok, secret} <- Helpers.get_user_secret(conn, tenant.auth_query, db_user) do
+        t = if secret.digest == :md5, do: :auth_query_md5, else: :auth_query
+        {:ok, {t, fn -> Map.put(secret, :alias, user.db_user_alias) end}}
+      end
+
+    GenServer.stop(conn, :normal, 5_000)
+    Logger.info("ProxyClient: Get secrets finished")
+    resp
+  end
+
+  @spec exchange_first(:password | :auth_query, fun(), binary(), binary(), binary()) ::
+          {binary(), map()}
+  def exchange_first(:password, secret, nonce, user, channel) do
+    message = Server.exchange_first_message(nonce)
+    server_first_parts = Helpers.parse_server_first(message, nonce)
+
+    {client_final_message, server_proof} =
+      Helpers.get_client_final(
+        :password,
+        secret.().password,
+        server_first_parts,
+        nonce,
+        user,
+        channel
+      )
+
+    sings = %{
+      client: List.last(client_final_message),
+      server: server_proof
+    }
+
+    {message, sings}
+  end
+
+  def exchange_first(:auth_query, secret, nonce, user, channel) do
+    secret = secret.()
+    message = Server.exchange_first_message(nonce, secret.salt)
+    server_first_parts = Helpers.parse_server_first(message, nonce)
+
+    sings =
+      Helpers.signatures(
+        secret.stored_key,
+        secret.server_key,
+        server_first_parts,
+        nonce,
+        user,
+        channel
+      )
+
+    {message, sings}
+  end
+
+  @spec timeout_check(atom, non_neg_integer) :: {:timeout, non_neg_integer, atom}
+  def timeout_check(key, timeout) do
+    {:timeout, timeout, key}
+  end
+
+  @spec handle_actions(map) :: [{:timeout, non_neg_integer, atom}]
+  defp handle_actions(%{} = data) do
+    heartbeat =
+      if data.heartbeat_interval > 0,
+        do: [{:timeout, data.heartbeat_interval, :heartbeat_check}],
+        else: []
+
+    idle =
+      if data.idle_timeout > 0, do: [{:timeout, data.idle_timeout, :idle_timeout}], else: []
+
+    idle ++ heartbeat
+  end
+
+  @spec app_name(any()) :: String.t()
+  def app_name(name) when is_binary(name) do
+    suffix = " via Supavisor"
+    # https://www.postgresql.org/docs/current/runtime-config-logging.html#GUC-APPLICATION-NAME
+    max_len = 64
+    suffix_len = 14
+
+    if String.length(name) <= max_len - suffix_len do
+      name <> suffix
+    else
+      truncated_name = String.slice(name, 0, max_len - suffix_len - 3)
```

Comment:
I assume the limit is in bytes, not in UTF8 codepoints. While perhaps no one will have UTF8 codepoints in their application name, better to be safe:

```suggestion
    if byte_size(name) <= max_len - suffix_len do
      name <> suffix
    else
      truncated_name = binary_slice(name, 0, max_len - suffix_len - 3)
```